flo-runner/src/main/java/com/spotify/flo/context/ForkingContext.java

```diff
@@ -0,0 +1,440 @@
+import com.spotify.flo.freezer.PersistingContext;
+import io.grpc.Context;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.PrintStream;
+import java.lang.management.ManagementFactory;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.time.LocalTime;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An {@link EvalContext} that runs tasks in sub-processes.
+ * <p>
+ * Forking can be disabled using the environment variable {@code FLO_DISABLE_FORKING=true}.
+ * <p>
+ * Forking is disabled by default when running in the debugger, but can be enabled by {@code FLO_FORCE_FORK=true}.
+ */
+class ForkingEvalContext implements EvalContext {
+
+  private static final Logger log = LoggerFactory.getLogger(ForwardingEvalContext.class);
+
+  private transient final EvalContext delegate;
+
+  private ForkingEvalContext(EvalContext delegate) {
+    this.delegate = Objects.requireNonNull(delegate);
+  }
+
+  static EvalContext composeWith(EvalContext baseContext) {
+    // Is the Java Debug Wire Protocol activated?
+    final boolean inDebugger = ManagementFactory.getRuntimeMXBean().
+        getInputArguments().stream().anyMatch(s -> s.contains("-agentlib:jdwp"));
+
+    final boolean disableForking = Boolean.parseBoolean(System.getenv("FLO_DISABLE_FORKING"));
+
+    final boolean forceFork = Boolean.parseBoolean(System.getenv("FLO_FORCE_FORK"));
+
+    if (forceFork) {
+      log.debug("Forking forcibly enabled (environment variable FLO_FORCE_FORK=true)");
+      return new ForkingEvalContext(baseContext);
+    } else if (disableForking) {
+      log.debug("Forking disabled (environment variable FLO_DISABLE_FORKING=true)");
+      return baseContext;
+    } else if (inDebugger) {
+      log.debug("In debugger, forking disabled (enable by setting environment variable FLO_FORCE_FORK=true)");
+      return baseContext;
+    } else {
+      return new ForkingEvalContext(baseContext);
+    }
+  }
+
+  @Override
+  public <T> Value<T> evaluateInternal(Task<T> task, EvalContext context) {
+    return delegate().evaluateInternal(task, context);
+  }
+
+  @Override
+  public <T> Value<T> value(Fn<T> value) {
+    return delegate().value(value);
+  }
+
+  @Override
+  public <T> Value<T> immediateValue(T value) {
+    return delegate().immediateValue(value);
+  }
+
+  @Override
+  public <T> Promise<T> promise() {
+    return delegate().promise();
+  }
+
+  @Override
+  public <T> Value<T> invokeProcessFn(TaskId taskId, Fn<Value<T>> processFn) {
+    if (delegate == null) {
```

Comment:
When would this happen?

Reply:
In the subprocess

Reply:
if a user uses the `processWithContext` method.

Reply:
yeah